### ISSUE TITLE
Repeatedly initializing and releasing player

### ISSUE THREAD

**Issue by xylik** [protyposis/ITEC-MediaPlayer#43]
Hi, am playing local mp4 files in recycler view and I've managed to break the player by quickly switching focus(playback) between two neighbour players in recycler view. By quickly I mean player came in focus and player initialization started but I would switch(release the player) to neighbour player before play(rendering frames) started, doing that repeatedly causes the player to silently crash with error log after several tries. 

I've created the log file [MediaPlayerCrashLog.rtf](https://drive.google.com/open?id=0Byn0xtAvVbCWMFRRZmJXS24wZjA) because it's not practical to paste such a huge log here. I suggest you to open it in text editor with wrap line set to false. In log file I've isolated 3 situations:
- last successful play just before first crash happens
- first player crash(player silently crashes with error log)
- second player crash(also silent crash) which is induced after continuing to switch focus between 2 neighbour players after player was already crashed with first player exception

Aditional notes:
- after player have crashed, it's not possible to play any video until i kill the app and restart it again.
- also when player crashed MediaPlayer.OnErrorListener() was not called.
- i have hardcoded in my app that after player release minimum 250ms must pass before new player can kick in with init/play. If time passed from player release until new play request is less than 250ms, init/play is delayed for diff = 250 - elapsed_time_since_release

Issue produced on Google nexus 5x.

**Comment by protyposis** [comment]
Hi, I can't really tell from the logs but it's probably an issue with the render surface, maybe because both players operate on the same recycled surface at the same time (e.g. player A is preparing with surface X while player B is still releasing surface X).

Can you prepare a small example activity so I can reproduce the issue?

**Comment by xylik** [comment]
I'll upload the code until day end.

**Comment by xylik** [comment]
Hi, sry for delay I was busy with other stuff and it wasn't trivial to create example. The example i uploaded is written in MVP and is actually my production code stripped of all unnecessary things. If you found your self difficult navigating through the code i can clean it even further or clarify some parts. There are lot of classes and interfaces but actually there is not much logic it's just MVP boilerplate ...

Classes you would be interested in:
MediaPlayerImp -> wrapper arround MediaPlayerExtended
MomentsView -> recyclerview which is responsible for hosting the video players
ItemVisibilityUtil -> util which calculates most visible item in recyclerview
MomentPresenterImp -> presenter responsible for controlling single video player in recyclerview, every item in recyclyer view has it's own presenter eg. if i have 10 items(players) in recyclerview I'll have 10 presenenters controlling them
MomentViewHolder -> viewholder(View) which represents single item in recyclerview
VideoPlayManager -> in my app am using 2 players and this class is responsible for picking the right player and assuring 250ms is passed between last player release and new player start. In the sample code am using only one player.

I've also put in assets folder several mp4 files and their starting thumbnails so you have everything setup for play.

https://github.com/xylik/MediaPlayerList

**Comment by protyposis** [comment]
Thanks for the example app and thank you again for helping me find an unpleasant bug!

There was a memory leak in `release()` that prevented codecs and audio tracks (and the PlaybackThread) from being released, ultimately leading to your observed errors once the maximum number of MediaCodec instances was reached. I assume this is what @piotrmadry observed as well and reported in #45.

I've already fixed it, but will do some more testing tomorrow before I release a new version.

**Comment by protyposis** [comment]
4.2.2-rc1 is online, please give it a try and let me know if it works for you. All leaks should be fixed and I can now scroll through the videos forever without playback failing.

In your sample app, I'm still getting two exceptions when switching many times quickly between videos, one from an invalid surface, and the other is a NullPointerException somewhere in your code. I'm not sure if the first exception is the player's fault or if your app is trying to use the same surface on two players at the same time.

**Comment by xylik** [comment]
I've tested example with new update and don't think my code is setting same surface on two players because on every play am releasing old player before I instantiate new one, but I will create new example with all MVP removed and playback counter rx thread removed just to eliminate code complexity and possible errors.

I've carefully examined the Verbose log and saw a pattern that potentially triggers the error state:
- **player release is being called before player would reach prepared state**("D/MediaPlayer: onPrepared") which allways resulted with Player writing to error log

- player at first is failing with this exception after release is being called but player didn't reach prepared state:
4143-5586/hr.fer.com.mplayerlist E/MediaCodecVideoDecoder: reinitCodec: illegal state
4143-5586/hr.fer.com.mplayerlist E/MediaPlayer: cannot create video decoder: null
4143-5586/hr.fer.com.mplayerlist E/MediaPlayer: cannot create audio decoder: null
4143-5586/hr.fer.com.mplayerlist E/MediaPlayer: prepareAsync() failed: cannot decode stream(s)
     java.io.IOException: cannot decode any stream
     at net.protyposis.android.mediaplayer.MediaPlayer.prepareInternal(MediaPlayer.java:328)
     at net.protyposis.android.mediaplayer.MediaPlayer.access$500(MediaPlayer.java:36)
     at net.protyposis.android.mediaplayer.MediaPlayer$2.run(MediaPlayer.java:423)
     at java.lang.Thread.run(Thread.java:818)
4143-4143/hr.fer.com.mplayerlist E/MediaPlayer: Error (1,-1004)

- then it fails with several different errors but it's allways failing after release is being called and player have not yet reached prepared state:

*1st error*
4143-5608/hr.fer.com.mplayerlist D/MediaCodecVideoDecoder: reinitCodec 35ms
4143-5608/hr.fer.com.mplayerlist E/MediaPlayer: cannot create audio decoder: null

*2nd error*
/hr.fer.com.mplayerlist E/MediaPlayer: prepareAsync() failed: something is in a wrong state
   java.lang.IllegalStateException
     at android.media.MediaCodec.native_dequeueOutputBuffer(Native Method)
     at android.media.MediaCodec.dequeueOutputBuffer(MediaCodec.java:2379)
     at net.protyposis.android.mediaplayer.MediaCodecDecoder.dequeueDecodedFrame(MediaCodecDecoder.java:309)
     at net.protyposis.android.mediaplayer.Decoders.decodeFrame(Decoders.java:81)
     at net.protyposis.android.mediaplayer.MediaPlayer.prepareInternal(MediaPlayer.java:365)
     at net.protyposis.android.mediaplayer.MediaPlayer.access$500(MediaPlayer.java:36)
     at net.protyposis.android.mediaplayer.MediaPlayer$2.run(MediaPlayer.java:423)
     at java.lang.Thread.run(Thread.java:818)
/hr.fer.com.mplayerlist E/MediaPlayer: Error (1,0)

*3rd error*
/hr.fer.com.mplayerlist E/BufferQueueProducer: [SurfaceTexture-0-4143-15] connect(P): already connected (cur=3 req=3)
/hr.fer.com.mplayerlist I/MediaCodec: native window already connected. Assuming no change of surface
/hr.fer.com.mplayerlist E/MediaCodec: configure failed with err 0xffffffea, resetting...

**Comment by protyposis** [comment]
You are right, the problem was that `release()` didn't wait for `prepareAsync()` to finish, which in some situations led to these errors.

I'm now at a point where whatever I do, I can't provoke an exception any more. Very cool, I've always been looking forward to this :) Thanks again for your valuable help, rc2 will follow tomorrow :) I think a new example app won't be necessary any more. Except if you had a really simple example with minimal dependencies, that would make a nice addition to the demo app!

**Comment by protyposis** [comment]
4.2.2-rc2 is online, please give it a try!

**Comment by xylik** [comment]
I've tested rc2 and player is super super stable :), I couldn't break it no matter what I do - great job!

There is one more thing that is still bodering me and that's recyclerview lagging while player initialization is going on. 

- Is it possible that some part of player initializaiton(some heavy operation for very short time) is done on main thread and if so could it be moved to background thread just to achieve smooth UI rendering ?
- I've updated my example with 250ms delay(from player release to new play) removed and you'll see that fast scrolling without lag is now not possible.

I'll be glad to write small example with minimal dependendencies which demonstrates play in recyclerview.

**Comment by protyposis** [comment]
Yup this lagging is extremely annoying, but I don't think it is the player, at least not directly. `prepareAsync` does nothing on the main thread, my measurements on a N5X show an execution time < 1 millisec. What's taking some time is `setDataSource`, about 30 millisecs, but moving it into the background thread doesn't fix the lagging even a bit. Then there's no other "heavy" methods left.

**Comment by protyposis** [comment]
Figured it out. It's actually the `release()` method that takes a long time (~200ms, because it now waits until prepareAsync has finished). If I put it into the background the lagging is completely gone, there's two problems though:
 * I can't move it into the background, because it would break MediaPlayer API compatibility,
 * and you shouldn't do it in your app because it again leads to exceptions regarding the surface (tried it with an AsyncTask).

So I think the best approach is the one you already had, add a delay and only call prepareAsync when the user stays at the video.

**Comment by xylik** [comment]
I see, the problem is actually more visible when user is slowly swypeing through list because when he is swypeing slowly to next player release is happening on earlier player and lag occures in transition to new player. When user do fast swipe, items(players) are passing to fast over the screen and neither one player init/release cycle happens due to 250ms trashold and everything is smooth.

Could it be somehow possible to create releaseAsync() - extend api for the sake of UX and somehow deal with surface destroy, am open to any approach to make it work because lag is really annoying and kills UX?

The problem is I can bypass issue when user do fast swipe but regular slow swipe I cant bypass with delay approach.

**Comment by protyposis** [comment]
So, I've been spending a lot of time on this issue and I was able to write a working `releaseAsync` method, but I won't publish it... it didn't solve the UI lag because the next problem which appeared then is the `setDataSource` method which also takes a while. So in the end, there are this 3 main methods in the player's lifecycle that take some time: `setDataSource`, `prepare`/`prepareAsync` and `release`, and I think it should not be the focus of this library to get them away from the main thread, as it complicates the architecture and largely deviates from the original MediaPlayer API.

In special cases like yours, I think you'll have to take care about getting the MediaPlayer off the main thread by yourself. I made a quick and dirty hack in your sample app and put the heavy MediaPlayer methods into AsyncTasks, and got buttery-smooth scrolling, but unfortunately also some exceptions and thread leaks that I couldn't manage to fix. It demonstrated though that your goal is reachable, that the result can be great, but also that AsyncTasks are not the way to go. I think a nice way might be managing the MediaPlayer on a `HandlerThread` and calling all methods from there (use the synchronized `prepare`).

I'm closing this issue now because the original issue, the exceptions and crashes (and associated leaks), are solved. If you want to further discuss how to get the player off the UI thread to get smooth scrolling, you're welcome to open another issue!
